xdg_shell.c

```diff
@@ -184,12 +184,38 @@ handle_xdg_shell_surface_unmap(struct wl_listener *listener, void *data)
 	view_unmap(view);
 }
 
+static void
+handle_xdg_shell_surface_commit(struct wl_listener *listener, void *data)
+{
+	struct cg_xdg_shell_view *xdg_shell_view = wl_container_of(listener, xdg_shell_view, commit);
+	struct wlr_xdg_toplevel_state *current_state = &xdg_shell_view->xdg_toplevel->current;
+
+	if (!current_state->maximized) {
+		/* Only check maximized surfaces */
+		return;
+	}
+
+	struct wlr_box geom;
+	wlr_xdg_surface_get_geometry(xdg_shell_view->xdg_toplevel->base, &geom);
+	if (current_state->width != geom.width || current_state->height != geom.height) {
+		/* The client changed its surface size in this commit even if expected to be maximized,
+		 * force recenter the surface in this case. */
+		wlr_xdg_toplevel_set_size(xdg_shell_view->xdg_toplevel, geom.width, geom.height);
+		wlr_xdg_toplevel_set_maximized(xdg_shell_view->xdg_toplevel, false);
```

Comment:
Why mark the client as unmaximized and why ask the client to resize itself to its current size?

Reply:
You're right it seems a bit unnecessary and seems to work well without these lines.